modules/aac/aac.c

```diff
@@ -1,38 +1,299 @@
+	(void)conf_get_u32(conf, "aac_aot",
+	                   &aac_aot);
+	(void)conf_get_u32(conf, "aac_bitrate",
+	                   &aac_bitrate);
+	(void)conf_get_u32(conf, "aac_constantduration",
+	                   &aac_constantduration);
+
+	if (aac_channels < 1 || aac_channels > 2) {
+		aac_channels = 2;
+	}
+
+	aac.ch = aac_channels;
+	aac.pch = aac_channels;
+
+	switch (aac_samplerate) {
+		case 8000:
+		case 11025:
+		case 12000:
+		case 16000:
+		case 22050:
+		case 24000:
+		case 32000:
+		case 44100:
+		case 48000:
+		case 64000:
+		case 88200:
+		case 96000:
+			break;
+		default:
+			aac_samplerate = 48000;
+			break;
+	}
+
+	aac.srate = aac_samplerate;
+	aac.crate = aac_samplerate;
+
+	switch (aac_aot) {
+	case AOT_AAC_LC:
+		/*  2: MPEG-4 AAC Low Complexity */
+		aac_profile = HIGH_QUALITY_AUDIO_PROFILE;
+		aac_constantduration = 1024;
+		aac_ratio = 1;
+		break;
+	case AOT_SBR:
+		/*  5: MPEG-4 AAC Low Complexity with Spectral Band
+		       Replication (HE-AAC) */
+		aac_profile = HIGH_EFFICIENCY_AAC_PROFILE;
+		aac_constantduration = 2048;
+		aac_ratio = 2;
+		break;
+	case AOT_PS: /* Stereo only! */
+		/* 29: MPEG-4 AAC Low Complexity with Spectral Band
+		       Replication and Parametric Stereo (HE-AAC v2) */
+		aac_profile = HIGH_EFFICIENCY_AAC_V2_PROFILE;
+		aac_constantduration = 2048;
+		aac_ratio = 2;
+		aac_channels = 2;
+		break;
+	case AOT_ER_AAC_LD:
+		/* 23: MPEG-4 AAC Low-Delay */
+		aac_profile = LOW_DELAY_AUDIO_PROFILE;
+		aac_ratio = 1;
+		if (aac_constantduration != 480 &&
+		    aac_constantduration != 512)
+			aac_constantduration = 480;
+	case AOT_ER_AAC_ELD:
+		/* 39: MPEG-4 AAC Enhanced Low-Delay */
+		aac_profile = ENHANCED_LOW_DELAY_AUDIO_PROFILE;
+		if (aac_channels == 2) ++aac_profile;
+		aac_ratio = 1;
+		switch (aac_constantduration) {
+		case 120:
+		case 128:
+		case 240:
+		case 256:
+		case 480:
+		case 512:
+			break;
+		default:
+			aac_constantduration = 120;
+			break;
+		}
+		break;
+	default:
+		warning("AAC Audio object types 2 (AAC-LC), 5 "
+			"(HE-AAC), 29 (HE-AAC v2), 23 (AAC-LD) "
+			"and 39 (AAC-ELD) are allowed.\n");
+		return EINVAL;
+	}
+
+	error = aacEncOpen(&enc, 0, 0);
+	if (error != AACENC_OK) {
+		warning("aac: Unable to open the encoder (0x%x)\n",
+			error);
+		return ENOMEM;
+	}
+
+	/* set mandatory encoder params: */
+	error |= aacEncoder_SetParam(enc, AACENC_AOT, aac_aot);
+	error |= aacEncoder_SetParam(enc, AACENC_SAMPLERATE, aac.srate);
+	error |= aacEncoder_SetParam(enc, AACENC_CHANNELMODE, aac.ch);
+	error |= aacEncoder_SetParam(enc, AACENC_BITRATE, aac_bitrate);
+	error |= aacEncoder_SetParam(enc, AACENC_TRANSMUX, TT_MP4_RAW);
+	/* set object specific encoder params: */
+	error |= aacEncoder_SetParam(enc, AACENC_GRANULE_LENGTH,
+	                             aac_constantduration/aac_ratio);
+	error |= aacEncEncode(enc, NULL, NULL, NULL, NULL);
+	error |= aacEncInfo(enc, &enc_info);
+	if (error != AACENC_OK) {
+		return EINVAL;
+	}
```

Comment:
this will leak memory. better to goto out and make sure "aacEncClose()" is called.

Reply:
Are you sure?
There is no dynamic memory allocation in this case before.
This was taken from the example code at
https://github.com/mstorsjo/fdk-aac/blob/d47c2c625940d2bd62b0cd581605c5122c3d2cfd/aac-enc.c#L117

But I am not an expert in such questions.